fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java

```diff
@@ -0,0 +1,201 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.rewrite.logical;
+
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.EqualTo;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.visitor.SlotExtractor;
+import org.apache.doris.nereids.trees.plans.JoinType;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.visitor.PlanVisitor;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Try to eliminate cross join via finding join conditions in filters and change the join orders.
+ * <p>
+ * <pre>
+ * For example:
+ *
+ * input:
+ * SELECT * FROM t1, t2, t3 WHERE t1.id=t3.id AND t2.id=t3.id
+ *
+ * output:
+ * SELECT * FROM t1 JOIN t3 ON t1.id=t3.id JOIN t2 ON t2.id=t3.id
+ * </pre>
+ * </p>
+ * TODO: This is tested by SSB queries currently, add more `unit` test for this rule
+ * when we have a plan building and comparing framework.
+ */
+public class ReorderJoin extends OneRewriteRuleFactory {
+    @Override
+    public Rule<Plan> build() {
+        return logicalFilter(subTree(LogicalJoin.class, LogicalFilter.class)).then(filter -> {
+            PlanCollector collector = new PlanCollector();
+            filter.accept(collector, null);
+            List<Plan> joinInputs = collector.joinInputs;
+            List<Expression> conjuncts = collector.conjuncts;
+
+            if (joinInputs.size() >= 3 && !conjuncts.isEmpty()) {
+                return reorderJoins(joinInputs, conjuncts);
```

Comment:
i think we chould add other reorder algorithm and use a config to decide witch algorithm should use here later. am i right?

Reply:
Yes, any candidate config name? BTW,  I prefer a unified config name prefix in the new planner, such `enable_nereids_xxx`.

Reply:
Added a session variable `enable_nereids_reorder_to_eliminate_cross_join`.